go-controller/pkg/util/gateway-init.go

```diff
@@ -159,15 +153,23 @@ func GatewayInit(clusterIPSubnet, nodeName, nicIP, physicalInterface,
 	var routerIP string
 	if routerMac == "" {
 		routerMac = GenerateMac()
+		err := lockNBForGateways()
```

Comment:
A more idiomatic way to do this that doesn't require sprinkling unlock calls around would be to put the stuiff that needs to be locked/unlocked in its own inline function where you can actually defer the unlock:

```
if routerMac == "" {
    routerMac = GenerateMac()
    if err := func() error {
                err := lockNBForGateways()
		if err != nil {
			return err
		}
                defer unlockNBForGateways()

		routerIP, err = generateGatewayIP()
		if err != nil {
			return err
		}

		stdout, stderr, err = RunOVNNbctl("--", "--may-exist", "lrp-add",
			gatewayRouter, "rtoj-"+gatewayRouter, routerMac, routerIP,
			"--", "set", "logical_router_port", "rtoj-"+gatewayRouter,
			"external_ids:connect_to_join=yes")
		if err != nil {
			return fmt.Errorf(Failed to add logical port to router, stdout: %q, "+
				"stderr: %q, error: %v", stdout, stderr, err)
		}
                return nil
    }(); err != nil {
        logrus.Errorf(err.String())
        return err
    }
}
```

Reply:
Thanks Dan for the suggestion. I will resubmit the PR with the changes you have suggested.